src/callbacks/profiler.cpp

```diff
@@ -61,83 +65,124 @@ static void prof_region_begin(const char *s, int c) {
   ev.message.ascii = s; 
   nvtxRangePushEx(&ev);
 }
-static void prof_region_end(const char *s) {
-  El::GPUManager::SynchronizeDevice();
+static void prof_region_end(const char *, bool sync) {
+  if (sync) {
+    El::GPUManager::SynchronizeDevice();
+  }
   nvtxRangePop();
 }
 #else
-static void prof_region_begin(const char *s, int c) {
+static void prof_region_begin(const char *, int, bool) {
   return;
 }
-static void prof_region_end(const char *s) {
+static void prof_region_end(const char *, bool) {
   return;
 }
 #endif
 }
 
 namespace lbann {
 
+lbann_callback_profiler::lbann_callback_profiler(bool sync) :
+  lbann_callback(), m_sync(sync) {
+#ifdef LBANN_NVPROF
+  nvtxNameCudaStreamA(El::GPUManager::Stream(), "Hydrogen");
+#endif  
+}
+
 void lbann_callback_profiler::on_epoch_begin(model *m) {
-  prof_region_begin("epoch", colors[0]);
+  prof_region_begin(("epoch " + std::to_string(m->get_cur_epoch())).c_str(),
+                    colors[0], m_sync);
 }
 
 void lbann_callback_profiler::on_epoch_end(model *m) {
-  prof_region_end("epoch");
+  prof_region_end(("epoch " + std::to_string(m->get_cur_epoch())).c_str(),
+                  m_sync);
 }
 
 void lbann_callback_profiler::on_batch_begin(model *m) {
-  prof_region_begin("batch", colors[1]);
+  prof_region_begin(("batch " + std::to_string(m->get_cur_step())).c_str(),
+                    colors[1], m_sync);
 }
 
 void lbann_callback_profiler::on_batch_end(model *m) {
-  prof_region_end("batch");  
+  prof_region_end(("batch " + std::to_string(m->get_cur_step())).c_str(),
+                  m_sync);
 }
 
 void lbann_callback_profiler::on_forward_prop_begin(model *m) {
-  prof_region_begin("forward", colors[2]);
+  prof_region_begin("forward", colors[2], m_sync);
 }
 
 void lbann_callback_profiler::on_forward_prop_end(model *m) {
-  prof_region_end("forward");
+  prof_region_end("forward", m_sync);
 }
 
 void lbann_callback_profiler::on_backward_prop_begin(model *m) {
-  prof_region_begin("backward", colors[3]);
+  prof_region_begin("backward", colors[3], m_sync);
 }
 
 void lbann_callback_profiler::on_backward_prop_end(model *m) {
-  prof_region_end("backward");
+  prof_region_end("backward", m_sync);
+}
+
+void lbann_callback_profiler::on_optimize_begin(model *m) {
+  prof_region_begin("optimize", colors[4], m_sync);
+}
+
+void lbann_callback_profiler::on_optimize_end(model *m) {
+  prof_region_end("optimize", m_sync);
 }
 
 int lbann_callback_profiler::get_color(Layer *l) {
   const std::string &lname = l->get_type();
-  int idx = 4;
-  if (lname == "fully_connected") {
-    idx = 5;
-  } else if (lname == "convolution") {
+  int idx = 5;
+  if (lname == "fully connected") {
```

Comment:
I don't think we should hardcode a colorscheme explicitly for Resnet-50. If we want all layers of the same type to have the same color, perhaps we can hash the layer type name.

Reply:
This wasn't really intended for ResNet-50 so much as just emphasizing commonly-used layers and layers that tend to be computationally expensive. (Admittedly, those and ResNet-50 tend to coincide...)